File: src/main/java/com/enterprise/iqk/evaluation/EvaluationScorer.java
package com.enterprise.iqk.evaluation;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.util.List;
import java.util.Locale;

@Component
@RequiredArgsConstructor
public class EvaluationScorer {
    private final ObjectMapper objectMapper;

    public CaseScores scoreCase(EvalCaseRecord evalCase,
                                String answer,
                                List<String> citations,
                                List<String> evidence,
                                boolean failed) {
        List<String> expectedKeywords = readJsonList(evalCase.getExpectedKeywordsJson());
        List<String> expectedCitations = readJsonList(evalCase.getExpectedCitationsJson());
        List<String> forbiddenKeywords = readJsonList(evalCase.getForbiddenKeywordsJson());
        String answerPool = (emptyIfBlank(answer) + "\n" + String.join("\n", evidence)).toLowerCase(Locale.ROOT);
        String citationPool = String.join("\n", citations).toLowerCase(Locale.ROOT);

        double keywordScore = expectedKeywords.isEmpty()
                ? (StringUtils.hasText(answer) ? 1.0 : 0.0)
                : hitRate(expectedKeywords, answerPool);
        double citationCoverage = expectedCitations.isEmpty()
                ? 1.0
                : hitRate(expectedCitations, citationPool);
        boolean forbiddenHit = forbiddenKeywords.stream()
                .filter(StringUtils::hasText)
                .anyMatch(keyword -> answerPool.contains(keyword.toLowerCase(Locale.ROOT)));
        double retrievalHit = expectedCitations.isEmpty()
                ? (!citations.isEmpty() || !evidence.isEmpty() || keywordScore > 0 ? 1.0 : 0.0)
                : (citationCoverage > 0 ? 1.0 : 0.0);
        double faithfulness = failed ? 0.0 : scoreFaithfulness(answer, citations);
        if (forbiddenHit) {
            keywordScore = 0.0;
            faithfulness = Math.min(faithfulness, 0.2);
        }

        double score = round(0.30 * retrievalHit
                + 0.25 * citationCoverage
                + 0.25 * keywordScore
                + 0.20 * faithfulness);
        return new CaseScores(round(retrievalHit), round(citationCoverage), round(keywordScore), round(faithfulness), score);
    }

    private double hitRate(List<String> expected, String actualLower) {
        if (expected == null || expected.isEmpty()) {
            return 1.0;
        }
        long hits = expected.stream()
                .filter(StringUtils::hasText)
                .filter(item -> actualLower.contains(item.toLowerCase(Locale.ROOT)))
                .count();
        return round(hits / (double) expected.size());
    }

    private double scoreFaithfulness(String answer, List<String> citations) {
        if (!StringUtils.hasText(answer)) {
            return 0.0;
        }
        if (citations == null || citations.isEmpty()) {
            return 0.5;
        }
        int markers = 0;
        for (int i = 1; i <= citations.size(); i++) {
            if (answer.contains("[" + i + "]")) {
                markers++;
            }
        }
        return round(Math.min(1.0, markers / (double) citations.size()));
    }

    private List<String> readJsonList(String json) {
        if (!StringUtils.hasText(json)) {
            return List.of();
        }
        try {
            return objectMapper.readValue(json, new TypeReference<List<String>>() {
            });
        } catch (JsonProcessingException ex) {
            return List.of();
        }
    }

    private double round(double value) {
        return Math.round(value * 10000.0) / 10000.0;
    }

    private String emptyIfBlank(String value) {
        return StringUtils.hasText(value) ? value : "";
    }

    public record CaseScores(double retrievalHit,
                             double citationCoverage,
                             double keywordScore,
                             double answerFaithfulness,
                             double score) {
    }
}

File: typescript/apps/api/src/evaluation/evaluation.scorer.ts
import { Injectable } from "@nestjs/common";

export interface EvaluationCaseInput {
  expectedKeywords?: unknown;
  expectedCitations?: unknown;
  forbiddenKeywords?: unknown;
}

export interface CaseScores {
  retrievalHit: number;
  citationCoverage: number;
  keywordScore: number;
  answerFaithfulness: number;
  score: number;
}

@Injectable()
export class EvaluationScorer {
  scoreCase(
    evalCase: EvaluationCaseInput,
    answer: string,
    citations: string[],
    evidence: string[],
    failed: boolean
  ): CaseScores {
    const expectedKeywords = stringList(evalCase.expectedKeywords);
    const expectedCitations = stringList(evalCase.expectedCitations);
    const forbiddenKeywords = stringList(evalCase.forbiddenKeywords);
    const answerPool = `${nonBlankOrEmpty(answer)}\n${evidence.join("\n")}`.toLowerCase();
    const citationPool = citations.join("\n").toLowerCase();

    let keywordScore = expectedKeywords.length === 0
      ? (answer.trim() ? 1 : 0)
      : hitRate(expectedKeywords, answerPool);
    const citationCoverage = expectedCitations.length === 0
      ? 1
      : hitRate(expectedCitations, citationPool);
    const forbiddenHit = forbiddenKeywords
      .filter((keyword) => keyword.trim())
      .some((keyword) => answerPool.includes(keyword.toLowerCase()));
    const retrievalHit = expectedCitations.length === 0
      ? (citations.length > 0 || evidence.length > 0 || keywordScore > 0 ? 1 : 0)
      : (citationCoverage > 0 ? 1 : 0);
    let answerFaithfulness = failed ? 0 : scoreFaithfulness(answer, citations);

    if (forbiddenHit) {
      keywordScore = 0;
      answerFaithfulness = Math.min(answerFaithfulness, 0.2);
    }

    const score = round(
      0.30 * retrievalHit
      + 0.25 * citationCoverage
      + 0.25 * keywordScore
      + 0.20 * answerFaithfulness
    );
    return {
      retrievalHit: round(retrievalHit),
      citationCoverage: round(citationCoverage),
      keywordScore: round(keywordScore),
      answerFaithfulness: round(answerFaithfulness),
      score
    };
  }
}

function stringList(value: unknown): string[] {
  return Array.isArray(value)
    ? value.map((item) => typeof item === "string" ? item : "")
    : [];
}

function hitRate(expected: string[], actualLower: string): number {
  if (expected.length === 0) {
    return 1;
  }
  const hits = expected
    .filter((item) => item.trim())
    .filter((item) => actualLower.includes(item.toLowerCase()))
    .length;
  return round(hits / expected.length);
}

function scoreFaithfulness(answer: string, citations: string[]): number {
  if (!answer.trim()) {
    return 0;
  }
  if (citations.length === 0) {
    return 0.5;
  }
  let markers = 0;
  for (let index = 1; index <= citations.length; index += 1) {
    if (answer.includes(`[${index}]`)) {
      markers += 1;
    }
  }
  return round(Math.min(1, markers / citations.length));
}

function nonBlankOrEmpty(value: string): string {
  return value.trim() ? value : "";
}

export function roundEvaluation(value: number): number {
  return round(value);
}

function round(value: number): number {
  return Math.round(value * 10000) / 10000;
}
